src/clojure/nrepl/util/completion.clj

```diff
@@ -8,8 +8,93 @@
   (:require [clojure.main])
   (:import [java.util.jar JarFile]
            [java.io File]
-           [java.lang.reflect Member]
-           [java.util.jar JarEntry]))
+           [java.lang.reflect Field Member]
+           [java.util.jar JarEntry]
+           [java.util.concurrent ConcurrentHashMap]))
+
+;; Code adapted from Compliment (https://github.com/alexander-yakushev/compliment)
+
+(defn annotate-keyword
+  [kw]
+  {:candidate kw :type :keyword})
+
+(defn all-keywords
+  []
+  (let [^Field field (.getDeclaredField clojure.lang.Keyword "table")]
+    (.setAccessible field true)
+    (.keySet ^ConcurrentHashMap (.get field nil))))
+
+(defn- resolve-namespace
+  [sym ns]
+  (get (ns-aliases ns) sym (find-ns sym)))
+
+(defn qualified-auto-resolved-keywords
+  "Given a namespace alias, a prefix, and a namespace, return completion
+  candidates for qualified, auto-resolved keywords (e.g. ::foo/bar)."
+  [ns-alias prefix ns]
+  (let [ns-alias-name (str (resolve-namespace (symbol ns-alias) ns))]
+    (sequence
+     (comp
+      (filter #(= (namespace %) ns-alias-name))
+      (filter #(.startsWith (name %) prefix))
+      (map #(str "::" ns-alias "/" (name %)))
+      (map annotate-keyword))
+     (all-keywords))))
+
+(defn unqualified-auto-resolved-keywords
+  "Given a prefix and a namespace, return completion candidates for
+  keywords that belong to the given namespace."
+  [prefix ns]
+  (sequence
+   (comp
+    (filter #(= (namespace %) (str ns)))
+    (filter #(.startsWith (name %) (subs prefix 2)))
+    (map #(str "::" (name %)))
+    (map annotate-keyword))
+   (all-keywords)))
+
+(defn keyword-namespace-aliases
+  "Given a prefix and a namespace, return completion candidates for namespace
+  aliases as auto-resolved keywords."
+  [prefix ns]
+  (sequence
+   (comp
+    (map (comp name first))
+    (filter (fn [^String alias-name] (.startsWith alias-name (subs prefix 2))))
+    (map #(str "::" (name %)))
+    (map annotate-keyword))
+   (ns-aliases ns)))
+
+(defn single-colon-keywords
+  "Given a prefix, return completion candidates for keywords that are either
+  unqualified or qualified with a synthetic namespace."
+  [prefix]
+  (sequence
+   (comp
+    (filter #(.startsWith (str %) (subs prefix 1)))
+    (map #(str ":" %))
+    (map annotate-keyword))
+   (all-keywords)))
+
+(defn keyword-candidates
+  [^String prefix ns]
+  (assert (string? prefix))
+  (let [double-colon? (.startsWith prefix "::")
+        single-colon? (.startsWith prefix ":")
+        slash-pos (.indexOf prefix "/")]
+    (cond
+      (and double-colon? (pos? slash-pos))
+      (let [ns-alias (subs prefix 2 slash-pos)
```

Comment:
Compliment [uses a regex](https://github.com/alexander-yakushev/compliment/blob/a112bb85b748d4a34da89ee783e7c64d8183dab7/src/compliment/sources/keywords.clj#L42) for this. I figured that since we need to find the position of the slash anyway, why not just use `subs`.